azurerm/internal/services/postgres/postgresql_flexible_server_database_resource.go

```diff
@@ -0,0 +1,162 @@
+				ValidateFunc: validate.DatabaseCollation,
+			},
+		},
+	}
+}
+
+func resourcePostgresqlFlexibleServerDatabaseCreate(d *pluginsdk.ResourceData, meta interface{}) error {
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+	client := meta.(*clients.Client).Postgres.FlexibleServerDatabaseClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	serverId, err := parse.FlexibleServerID(d.Get("server_id").(string))
+	if err != nil {
+		return err
+	}
+
+	id := parse.NewFlexibleServerDatabaseID(subscriptionId, serverId.ResourceGroup, serverId.Name, name)
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, serverId.ResourceGroup, serverId.Name, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for present of existing %q: %+v", id, err)
+			}
+		}
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return tf.ImportAsExistsError("azurerm_postgresql_flexible_server_database", id.ID())
+		}
+	}
+
+	properties := postgresqlflexibleservers.Database{
+		DatabaseProperties: &postgresqlflexibleservers.DatabaseProperties{
+			Charset:   utils.String(d.Get("charset").(string)),
+			Collation: utils.String(d.Get("collation").(string)),
+		},
+	}
+
+	future, err := client.Create(ctx, id.ResourceGroup, id.FlexibleServerName, id.DatabaseName, properties)
+	if err != nil {
+		return fmt.Errorf("creating Azure Postgresql Flexible Server database %q on server %q (Resource Group %q): %+v", id.DatabaseName, id.FlexibleServerName, id.ResourceGroup, err)
+	}
+
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("waiting for creation of Azure Postgresql Flexible Server database %q on server %q (Resource Group %q): %+v", id.DatabaseName, id.FlexibleServerName, id.ResourceGroup, err)
+	}
+
+	d.SetId(id.ID())
+	return resourcePostgresqlFlexibleServerDatabaseRead(d, meta)
+}
+
+func resourcePostgresqlFlexibleServerDatabaseRead(d *pluginsdk.ResourceData, meta interface{}) error {
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
+	client := meta.(*clients.Client).Postgres.FlexibleServerDatabaseClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.FlexibleServerDatabaseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.FlexibleServerName, id.DatabaseName)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] Azure Postgresql Flexible Server database %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving %q: %+v", id, err)
+	}
+	d.Set("name", id.DatabaseName)
+	d.Set("server_id", parse.NewFlexibleServerID(subscriptionId, id.ResourceGroup, id.FlexibleServerName).ID())
+	d.Set("charset", resp.Charset)
+	d.Set("collation", resp.Collation)
```

Comment:
we'll need to pull these from the `properties` object since that's where they sit in the API:
```suggestion
	if props := resp.DatabaseProperties; props != nil {
		d.Set("charset", props.Charset)
		d.Set("collation", props.Collation)
	}
```

Reply:
changed